src/main/java/org/openpnp/spi/base/AbstractCamera.java

```diff
@@ -253,16 +296,172 @@ public Location getLocation(HeadMountable tool) {
+     * @param heightAboveCamera
+     * @return the Z coordinate that is the given height above the camera
+     */
+    protected Length getZCoordinateAboveCamera(Length heightAboveCamera) {
+        return heightAboveCamera.add(getActualLocation().getLengthZ());
+    }
+    
     @Override
     public Location getUnitsPerPixel() {
-        return unitsPerPixel;
+        return getUnitsPerPixel(defaultZ);
+    }
+
+    @Override
+    public Location getUnitsPerPixel(Length atHeight) {
+        if (atHeight == null) {
+            atHeight = defaultZ;
+        }
+        LengthUnit units = unitsPerPixel.getUnits();
+        Location uppCal1 = unitsPerPixel;
+        Location uppCal2 = unitsPerPixelSecondary.convertToUnits(units);
+        if (uppCal1.getZ() == uppCal2.getZ()) {
+            // Calibration wasn't performed at two different heights above the camera so just
+            // return the primary units per pixels
+            return unitsPerPixel;
+        }
+
+        double heightAboveCamera = getHeightAboveCamera(atHeight).convertToUnits(units).getValue();
+
+        // Linearly interpolate between the two calibration points
+        double k = (heightAboveCamera - uppCal2.getZ()) / (uppCal1.getZ() - uppCal2.getZ());
+        return new Location(units, k * (uppCal1.getX() - uppCal2.getX()) + uppCal2.getX(),
+                k * (uppCal1.getY() - uppCal2.getY()) + uppCal2.getY(), heightAboveCamera, 0.0);
+    }
+
+    @Override
+    public Location getUnitsPerPixel(Location location) {
+        if (location == null) {
+            return getUnitsPerPixel(defaultZ);
+        }
+        return getUnitsPerPixel(location.getLengthZ());
     }
 
     @Override
     public void setUnitsPerPixel(Location unitsPerPixel) {
         this.unitsPerPixel = unitsPerPixel;
     }
 
+    @Override
+    public Location getUnitsPerPixelSecondary() {
+        return unitsPerPixelSecondary;
+    }
+
+    @Override
+    public void setUnitsPerPixelSecondary(Location unitsPerPixelSecondary) {
+        this.unitsPerPixelSecondary = unitsPerPixelSecondary;
+    }
+
+    @Override
+    public Length getDefaultZ() {
+        return defaultZ;
+    }
+
+    @Override
+    public void setDefaultZ(Length defaultZ) {
+        this.defaultZ = defaultZ;
+    }
+
+    @Override
+    public Length estimateZCoordinateOfObject(Location observedUnitsPerPixel) throws Exception {
+        LengthUnit units = observedUnitsPerPixel.getUnits();
+        double uppX = Math.abs(observedUnitsPerPixel.getX());
+        double uppY = Math.abs(observedUnitsPerPixel.getY());
+
+        Location uppCal1 = unitsPerPixel.convertToUnits(units);
+        Location uppCal2 = unitsPerPixelSecondary.convertToUnits(units);
+
+        if (uppCal1.getZ() == uppCal2.getZ()) {
+            throw new Exception("Camera Units Per Pixel has not been calibrated at two different " +
+                    "heights.");
+        }
+        
+        if (!Double.isFinite(uppX) && !Double.isFinite(uppY)) {
```

Comment:
Should this be:
 if (!(Double.isFinite(uppX) && Double.isFinite(uppY))) {

Reply:
I don't think so - as long as we have one finite non-zero value we can calculate the height.  For example, if the machine is only moved in one axis the other axis may have a zero answer (or Nan if the pixels in that direction also didn't change).  The other case is the machine barely moved in one axis but the pixels in that direction didn't change so you get infinity for an answer on that axis.  I'll add a comment to the code to make that clear.

Reply:
Ah, I see it now. Thanks.

`if (!(Double.isFinite(uppX) || Double.isFinite(uppY))) {`

being the ideal ;-)